connection/connection.go

```diff
@@ -139,30 +142,36 @@ func WithOtelTracing() Option {
 }
 
 type options struct {
-	reconnect         func() bool
+	reconnect         func(context.Context) bool
 	timeout           time.Duration
 	metricsManager    metrics.CSIMetricsManager
 	enableOtelTracing bool
 }
 
 // connect is the internal implementation of Connect. It has more options to enable testing.
 func connect(
+	ctx context.Context,
 	address string,
 	connectOptions []Option) (*grpc.ClientConn, error) {
+	logger := klog.FromContext(ctx)
 	var o options
 	for _, option := range connectOptions {
 		option(&o)
 	}
 
+	bc := backoff.DefaultConfig
+	bc.MaxDelay = time.Second
 	dialOptions := []grpc.DialOption{
-		grpc.WithInsecure(),                    // Don't use TLS, it's usually local Unix domain socket in a container.
-		grpc.WithBackoffMaxDelay(time.Second),  // Retry every second after failure.
+		grpc.WithTransportCredentials(insecure.NewCredentials()), // Don't use TLS, it's usually local Unix domain socket in a container.
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: bc}),  // Retry every second after failure.
 		grpc.WithBlock(),                       // Block until connection succeeds.
 		grpc.WithIdleTimeout(time.Duration(0)), // Never close connection because of inactivity.
 	}
 
 	if o.timeout > 0 {
-		dialOptions = append(dialOptions, grpc.WithTimeout(o.timeout))
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, o.timeout)
+		defer cancel()
```

Comment:
Note to myself: this works because `connect` waits for dialing to finish or fail. Otherwise returning would cause an on-going dial attempt to be aborted.

Reply:
There is one problem that was already in the original code: as described in https://github.com/grpc/grpc-go/issues/133, `WithTimeout` also prevents all future attempts to reconnect. That might be surprising when someone wants to use `Connect` with a `OnConnectionLoss` callback that doesn't abort.

Can you add a comment to `WithTimeout` explaining that setting a timeout prevent reconnecting?

Reply:
added: https://github.com/kubernetes-csi/csi-lib-utils/pull/149/commits/de384e6e6810061fcb4541102822f8ca8a3cc03d#diff-b9adde9612439aa9df1ceebc36325654ae9d11a48c508b63995688809c1a1115R173-R175

Reply:
@pohly Sorry for my misunderstanding. I have removed the comment.

Regarding this:
> I meant adding the comment to `func WithTimeout(...)`. Then it becomes part of the API documentation.

I'm not fully understanding the implication, so I would like to clarify the meaning.
Is this suggesting that a comment should be added to the following page?
https://pkg.go.dev/context#WithTimeout

If my understanding above is correct, please let me know if there's anything else I should do to clarify or address this.

Reply:
No. I meant adding it to https://github.com/kubernetes-csi/csi-lib-utils/blob/f82f9de5b8aeb3c3b236d7f58fc5eeab34438078/connection/connection.go#L120-L125